File: android/src/main/kotlin/com/example/dengage_flutter/InAppInline.kt
package com.example.dengage_flutter

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.net.http.SslError
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.view.ViewTreeObserver
import android.webkit.ConsoleMessage
import android.webkit.SslErrorHandler
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import com.dengage.sdk.Dengage
import com.dengage.sdk.ui.inappmessage.InAppInlineElement
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

private const val TAG = "den/InAppInline"

class InAppInline internal constructor(
    context: Context,
    creationParams: HashMap<String, Any>,
    activity: Activity,
    messenger: BinaryMessenger,
    viewId: Int
) :
    PlatformView {
    private lateinit var inAppInlineElement: InAppInlineElement
    private var methodChannel: MethodChannel
    private var isContentLoaded = false
    private var isNotFoundNotified = false
    private var isInitialized = false
    private val mainHandler = Handler(Looper.getMainLooper())
    private var visibilityListener: ViewTreeObserver.OnGlobalLayoutListener? = null
    private var notFoundRunnable: Runnable? = null

    override fun getView(): View {
        return inAppInlineElement
    }

    init {
        Log.d(TAG, "========== InAppInline INIT START ==========")
        Log.d(TAG, "viewId: $viewId")
        Log.d(TAG, "creationParams: $creationParams")
        
        methodChannel = MethodChannel(messenger, "plugins.dengage/inappinline_$viewId")
        Log.d(TAG, "MethodChannel created: plugins.dengage/inappinline_$viewId")

        inAppInlineElement = InAppInlineElement(context)
        Log.d(TAG, "InAppInlineElement created")
        
        inAppInlineElement.visibility = View.INVISIBLE
        inAppInlineElement.alpha = 0f
        Log.d(TAG, "Initial visibility: INVISIBLE, alpha: 0")

        val propertyId = creationParams["propertyId"] as String
        val customParams = creationParams["customParams"] as HashMap<String, String>?
        val screenName = creationParams["screenName"] as String?
        
        Log.d(TAG, "Calling Dengage.showInlineInApp with:")
        Log.d(TAG, "  - propertyId: $propertyId")
        Log.d(TAG, "  - screenName: $screenName")
        Log.d(TAG, "  - customParams: $customParams")
        
        Dengage.showInlineInApp(
            activity = activity,
            propertyId = propertyId,
            inAppInlineElement = inAppInlineElement,
            customParams = customParams,
            screenName = screenName
        )
        Log.d(TAG, "Dengage.showInlineInApp called")
        
        mainHandler.postDelayed({
            setupWebViewClient()
            setupWebChromeClient()
            setupVisibilityListener()
            isInitialized = true
            Log.d(TAG, "Initialization completed (delayed)")
            
            logWebViewState()
        }, 100)
        
        Log.d(TAG, "========== InAppInline INIT END ==========")
    }

    private fun logWebViewState() {
        Log.d(TAG, "========== WebView State ==========")
        Log.d(TAG, "URL: ${inAppInlineElement.url}")
        Log.d(TAG, "Original URL: ${inAppInlineElement.originalUrl}")
        Log.d(TAG, "Title: ${inAppInlineElement.title}")
        Log.d(TAG, "Progress: ${inAppInlineElement.progress}%")
        Log.d(TAG, "Visibility: ${inAppInlineElement.visibility}")
        Log.d(TAG, "Width: ${inAppInlineElement.width}, Height: ${inAppInlineElement.height}")
        Log.d(TAG, "Settings JS Enabled: ${inAppInlineElement.settings.javaScriptEnabled}")
        Log.d(TAG, "===================================")
    }

    private fun setupWebChromeClient() {
        Log.d(TAG, "Setting up WebChromeClient")
        
        inAppInlineElement.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                super.onProgressChanged(view, newProgress)
                Log.d(TAG, ">>> onProgressChanged - Progress: $newProgress%")
                
                if (newProgress == 100) {
                    Log.d(TAG, ">>> onProgressChanged - Loading complete!")
                }
            }
            
            override fun onConsoleMessage(consoleMessage: ConsoleMessage?): Boolean {
                Log.d(TAG, ">>> Console [${consoleMessage?.messageLevel()}]: ${consoleMessage?.message()}")
                Log.d(TAG, ">>> Console - Source: ${consoleMessage?.sourceId()}:${consoleMessage?.lineNumber()}")
                return super.onConsoleMessage(consoleMessage)
            }
            
            override fun onReceivedTitle(view: WebView?, title: String?) {
                super.onReceivedTitle(view, title)
                Log.d(TAG, ">>> onReceivedTitle - Title: $title")
            }
        }
        Log.d(TAG, "WebChromeClient set")
    }

    private fun setupWebViewClient() {
        Log.d(TAG, "Setting up WebViewClient")
        
        val existingClient = inAppInlineElement.webViewClient
        Log.d(TAG, "Existing WebViewClient: $existingClient")
        
        inAppInlineElement.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                existingClient?.onPageStarted(view, url, favicon)
                Log.d(TAG, ">>> onPageStarted - URL: $url")
                Log.d(TAG, ">>> onPageStarted - isContentLoaded: $isContentLoaded, isNotFoundNotified: $isNotFoundNotified")
                
                cancelNotFoundCheck()
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                existingClient?.onPageFinished(view, url)
                Log.d(TAG, ">>> onPageFinished - URL: $url")
                Log.d(TAG, ">>> onPageFinished - isContentLoaded: $isContentLoaded, isNotFoundNotified: $isNotFoundNotified")
                Log.d(TAG, ">>> onPageFinished - view: $view")
                
                if (!isContentLoaded && !isNotFoundNotified && view != null) {
                    val currentUrl = view.url
                    Log.d(TAG, ">>> onPageFinished - currentUrl: $currentUrl")
                    
                    view.evaluateJavascript(
                        "(function() { return document.body ? document.body.innerHTML.trim().length : 0; })()"
                    ) { result ->
                        val contentLength = result?.replace("\"", "")?.toIntOrNull() ?: 0
                        Log.d(TAG, ">>> onPageFinished - Body content length: $contentLength")
                        
                        if (contentLength > 0) {
                            Log.d(TAG, ">>> onPageFinished - Content found! Setting isContentLoaded = true")
                            
                            if (!isContentLoaded && !isNotFoundNotified) {
                                isContentLoaded = true
                                cancelNotFoundCheck()
                                
                                inAppInlineElement.visibility = View.VISIBLE
                                inAppInlineElement.animate()
                                    .alpha(1f)
                                    .setDuration(150)
                                    .start()
                                Log.d(TAG, ">>> onPageFinished - View made VISIBLE with animation")
                                
                                collectWebViewInfoAndNotify(view)
                            }
                        } else {
                            Log.d(TAG, ">>> onPageFinished - No content, scheduling notFound check")
                            scheduleNotFoundCheck(1000)
                        }
                    }
                    return
                } else {
                    Log.d(TAG, ">>> onPageFinished - SKIPPED: isContentLoaded=$isContentLoaded, isNotFoundNotified=$isNotFoundNotified, view=$view")
                }
            }

            override fun onReceivedError(
                view: WebView?,
                errorCode: Int,
                description: String?,
                failingUrl: String?
            ) {
                existingClient?.onReceivedError(view, errorCode, description, failingUrl)
                Log.e(TAG, ">>> onReceivedError - errorCode: $errorCode, description: $description, failingUrl: $failingUrl")
                
                try {
                    methodChannel.invokeMethod("onContentError", mapOf(
                        "errorCode" to errorCode,
                        "description" to description
                    ))
                    Log.d(TAG, ">>> onReceivedError - Flutter'a onContentError gönderildi")
                } catch (e: Exception) {
                    Log.e(TAG, ">>> onReceivedError - Flutter invoke error: ${e.message}")
                }
            }

            override fun onReceivedError(
                view: WebView?,
                request: WebResourceRequest?,
                error: WebResourceError?
            ) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    Log.e(TAG, ">>> onReceivedError (API 23+)")
                    Log.e(TAG, ">>>   URL: ${request?.url}")
                    Log.e(TAG, ">>>   isForMainFrame: ${request?.isForMainFrame}")
                    Log.e(TAG, ">>>   errorCode: ${error?.errorCode}")
                    Log.e(TAG, ">>>   description: ${error?.description}")
                }
                super.onReceivedError(view, request, error)
            }
            
            override fun onReceivedHttpError(
                view: WebView?,
                request: WebResourceRequest?,
                errorResponse: WebResourceResponse?
            ) {
                Log.e(TAG, ">>> onReceivedHttpError")
                Log.e(TAG, ">>>   URL: ${request?.url}")
                Log.e(TAG, ">>>   Status Code: ${errorResponse?.statusCode}")
                Log.e(TAG, ">>>   Reason: ${errorResponse?.reasonPhrase}")
                Log.e(TAG, ">>>   Headers: ${errorResponse?.responseHeaders}")
                super.onReceivedHttpError(view, request, errorResponse)
            }

            override fun onReceivedSslError(
                view: WebView?,
                handler: SslErrorHandler?,
                error: SslError?
            ) {
                Log.e(TAG, ">>> onReceivedSslError - error: $error")
                super.onReceivedSslError(view, handler, error)
            }
            
            override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                Log.d(TAG, ">>> shouldOverrideUrlLoading - URL: $url")
                return existingClient?.shouldOverrideUrlLoading(view, url) ?: super.shouldOverrideUrlLoading(view, url)
            }

            override fun shouldOverrideUrlLoading(
                view: WebView?,
                request: WebResourceRequest?
            ): Boolean {
                Log.d(TAG, ">>> shouldOverrideUrlLoading (API 24+)")
                Log.d(TAG, ">>>   URL: ${request?.url}")
                Log.d(TAG, ">>>   Method: ${request?.method}")
                Log.d(TAG, ">>>   Headers: ${request?.requestHeaders}")
                return super.shouldOverrideUrlLoading(view, request)
            }
            
            override fun onLoadResource(view: WebView?, url: String?) {
                existingClient?.onLoadResource(view, url)
                Log.d(TAG, ">>> onLoadResource - URL: $url")
            }

            override fun shouldInterceptRequest(
                view: WebView?,
                request: WebResourceRequest?
            ): WebResourceResponse? {
                Log.d(TAG, ">>> shouldInterceptRequest")
                Log.d(TAG, ">>>   URL: ${request?.url}")
                Log.d(TAG, ">>>   Method: ${request?.method}")
                Log.d(TAG, ">>>   Headers: ${request?.requestHeaders}")
                Log.d(TAG, ">>>   isForMainFrame: ${request?.isForMainFrame}")
                return super.shouldInterceptRequest(view, request)
            }

            override fun shouldInterceptRequest(view: WebView?, url: String?): WebResourceResponse? {
                Log.d(TAG, ">>> shouldInterceptRequest (legacy) - URL: $url")
                return super.shouldInterceptRequest(view, url)
            }
        }
        Log.d(TAG, "WebViewClient set (wrapped existing)")
    }

    private fun setupVisibilityListener() {
        Log.d(TAG, "Setting up visibility listener")
        var layoutChangeCount = 0
        
        visibilityListener = ViewTreeObserver.OnGlobalLayoutListener {
            if (!isInitialized) return@OnGlobalLayoutListener
            
            layoutChangeCount++
            val visibility = inAppInlineElement.visibility
            val visibilityStr = when(visibility) {
                View.VISIBLE -> "VISIBLE"
                View.INVISIBLE -> "INVISIBLE"
                View.GONE -> "GONE"
                else -> "UNKNOWN($visibility)"
            }
            val width = inAppInlineElement.width
            val height = inAppInlineElement.height
            
            if (layoutChangeCount <= 5 || layoutChangeCount % 10 == 0) {
                Log.d(TAG, ">>> onGlobalLayout #$layoutChangeCount - visibility: $visibilityStr, width: $width, height: $height")
                Log.d(TAG, ">>> onGlobalLayout - isContentLoaded: $isContentLoaded, isNotFoundNotified: $isNotFoundNotified")
            }
            if (isContentLoaded || isNotFoundNotified) return@OnGlobalLayoutListener
            
            if (visibility == View.GONE) {
                Log.d(TAG, ">>> onGlobalLayout - View is GONE, scheduling notFound check")
                scheduleNotFoundCheck(500)
            }
        }
        inAppInlineElement.viewTreeObserver.addOnGlobalLayoutListener(visibilityListener)
        Log.d(TAG, "Visibility listener added")
    }

    private fun scheduleNotFoundCheck(delayMs: Long) {

        cancelNotFoundCheck()
        
        Log.d(TAG, "Scheduling notFound check in ${delayMs}ms")
        notFoundRunnable = Runnable {
            if (!isContentLoaded && !isNotFoundNotified) {
                val visibility = inAppInlineElement.visibility
                val width = inAppInlineElement.width
                val height = inAppInlineElement.height
                
                Log.d(TAG, ">>> notFound check - visibility: $visibility, width: $width, height: $height")
                

                if (visibility == View.GONE) {
                    notifyContentNotFound()
                }
            }
        }
        mainHandler.postDelayed(notFoundRunnable!!, delayMs)
    }

    private fun cancelNotFoundCheck() {
        notFoundRunnable?.let {
            mainHandler.removeCallbacks(it)
            Log.d(TAG, "NotFound check cancelled")
        }
        notFoundRunnable = null
    }

    private fun notifyContentNotFound() {
        if (isNotFoundNotified) {
            Log.d(TAG, "notifyContentNotFound - Already notified, skipping")
            return
        }
        isNotFoundNotified = true
        
        Log.d(TAG, "========== CONTENT NOT FOUND ==========")
        Log.d(TAG, "View visibility: ${inAppInlineElement.visibility}")
        Log.d(TAG, "View width: ${inAppInlineElement.width}, height: ${inAppInlineElement.height}")
        Log.d(TAG, "WebView URL: ${inAppInlineElement.url}")
        Log.d(TAG, "WebView Original URL: ${inAppInlineElement.originalUrl}")
        
        try {
            methodChannel.invokeMethod("onContentNotFound", null)
            Log.d(TAG, "Flutter'a onContentNotFound gönderildi")
        } catch (e: Exception) {
            Log.e(TAG, "Error invoking onContentNotFound: ${e.message}")
        }
    }

    private fun collectWebViewInfoAndNotify(webView: WebView) {
        Log.d(TAG, "collectWebViewInfoAndNotify - Starting")
        Log.d(TAG, "WebView URL: ${webView.url}")
        Log.d(TAG, "WebView Title: ${webView.title}")
        
        webView.evaluateJavascript(
            "(function() { return document.body.scrollHeight; })();"
        ) { contentHeightStr ->
            Log.d(TAG, "JavaScript scrollHeight result: $contentHeightStr")
            val contentHeight = contentHeightStr?.replace("\"", "")?.toDoubleOrNull()
            
            webView.evaluateJavascript(
                "(function() { return document.body.scrollWidth; })();"
            ) { contentWidthStr ->
                Log.d(TAG, "JavaScript scrollWidth result: $contentWidthStr")
                val contentWidth = contentWidthStr?.replace("\"", "")?.toDoubleOrNull()
                
                mainHandler.post {
                    try {
                        val webViewInfo = mapOf(
                            "url" to webView.url,
                            "title" to webView.title,
                            "contentHeight" to contentHeight,
                            "contentWidth" to contentWidth,
                            "canGoBack" to webView.canGoBack(),
                            "canGoForward" to webView.canGoForward()
                        )
                        
                        Log.d(TAG, "========== CONTENT LOADED ==========")
                        Log.d(TAG, "WebView info: $webViewInfo")
                        methodChannel.invokeMethod("onContentLoaded", webViewInfo)
                        Log.d(TAG, "Flutter'a onContentLoaded gönderildi")
                    } catch (e: Exception) {
                        Log.e(TAG, "Error invoking onContentLoaded: ${e.message}")
                    }
                }
            }
        }
    }

    override fun dispose() {
        Log.d(TAG, "dispose() called")
        cancelNotFoundCheck()
        visibilityListener?.let {
            inAppInlineElement.viewTreeObserver.removeOnGlobalLayoutListener(it)
        }
        inAppInlineElement.destroy()
        Log.d(TAG, "InAppInlineElement destroyed")
    }
}

File: ios/Classes/InAppinline.swift
import Foundation
import Flutter
import UIKit
import WebKit
import Dengage

private let TAG = "[den/InAppInline]"

class InAppinline: NSObject, FlutterPlatformView {
    private var _nativeWebView: InAppInlineElementView
    private var methodChannel: FlutterMethodChannel
    private var isContentLoaded = false
    private var isNotFoundNotified = false
    private var isInitialized = false
    private weak var currentWebView: WKWebView?
    private var visibilityObserver: NSKeyValueObservation?
    private var boundsObserver: NSKeyValueObservation?
    
    // Store Dengage parameters
    private var propertyId: String?
    private var screenName: String?
    private var customParams: Dictionary<String, String>?
    private var dengageInitialized = false
    private let viewId: Int64
    
    // Polling timer for bounds checking (more reliable than KVO on iOS)
    private var boundsCheckTimer: Timer?
    private var lastKnownBounds: CGRect = .zero
    
    func view() -> UIView {
        return _nativeWebView
    }
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        messenger: FlutterBinaryMessenger
    ) {
     
        
        // Initialize viewId BEFORE super.init (required for let properties)
        self.viewId = viewId
        
        _nativeWebView = InAppInlineElementView(frame: frame)
        print("\(TAG) InAppInlineElementView created with frame: \(frame)")
        
        methodChannel = FlutterMethodChannel(
            name: "plugins.dengage/inappinline_\(viewId)",
            binaryMessenger: messenger
        )

        
        super.init()
        
        // CRITICAL: Set autoresizing mask so view resizes with parent
        _nativeWebView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        print("\(TAG) autoresizingMask set to flexibleWidth + flexibleHeight")
        
        // Start invisible, will animate to visible when content loads
        _nativeWebView.alpha = 0.0
        _nativeWebView.isHidden = false

        
        // Store Dengage parameters (will be used when view is laid out)
        if let data = args as? [String: Any] {
            self.propertyId = data["propertyId"] as? String
            self.customParams = data["customParams"] as? Dictionary<String, String>
            self.screenName = data["screenName"] as? String
            
           
        } else {
            print("\(TAG) ERROR: args is not [String: Any]!")
        }
        
        // Wait for Flutter to layout the view, then initialize Dengage
        // 0.3s is enough for Flutter's initial layout pass
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) { [weak self] in
            guard let self = self else { return }
            print("\(TAG) Initializing Dengage after layout delay")
            print("\(TAG) Current bounds: \(self._nativeWebView.bounds)")
            self.initializeDengage()
        }
        
        print("\(TAG) ========== InAppInline INIT END ==========")
    }
    
    deinit {
        print("\(TAG) deinit called - cleaning up observers")
        visibilityObserver?.invalidate()
        boundsObserver?.invalidate()
        visibilityObserver = nil
        boundsObserver = nil
        print("\(TAG) deinit completed")
    }
    
    private func logWebViewState() {
        guard let webView = findWKWebView(in: _nativeWebView) else {
            print("\(TAG) logWebViewState - WKWebView not found")
            return
        }
        
      
    }
    
    private func checkIfAlreadyLoaded() {
        guard let webView = findWKWebView(in: _nativeWebView) else {
            print("\(TAG) checkIfAlreadyLoaded - WKWebView not found, will wait for navigation events")
            return
        }
        
        print("\(TAG) checkIfAlreadyLoaded - isLoading: \(webView.isLoading), progress: \(webView.estimatedProgress), URL: \(webView.url?.absoluteString ?? "nil")")
        
        // If already loaded, check content
        if !webView.isLoading && webView.estimatedProgress >= 1.0 {
            print("\(TAG) Page already loaded, checking content...")
            checkContentAndNotify(webView: webView)
        } else {
            print("\(TAG) Page not loaded yet, will wait for didFinish navigation event")
        }
    }
    
    private func checkContentAndNotify(webView: WKWebView, retryCount: Int = 0) {
        let attemptMsg = "checkContentAndNotify attempt #\(retryCount + 1)"
        print("\(TAG) \(attemptMsg)")
        sendDebugLog(attemptMsg)
        
        // Check if view is hidden by SDK (means no content)
        if _nativeWebView.isHidden {
            let msg = "View is HIDDEN by SDK → content not found"
            print("\(TAG) \(msg)")
            sendDebugLog(msg)
            if !isContentLoaded && !isNotFoundNotified {
                notifyContentNotFound()
            }
            return
        }
        
        // HTML content check - is body empty?
        let script = "(function() { return document.body ? document.body.innerHTML.trim().length : 0; })()"
        
        webView.evaluateJavaScript(script) { [weak self] result, error in
            guard let self = self else { return }
            
            let contentLength = (result as? Int) ?? 0
            let lengthMsg = "Body content length: \(contentLength)"
            print("\(TAG) \(lengthMsg)")
            self.sendDebugLog(lengthMsg)
            
            if let error = error {
                print("\(TAG) checkContentAndNotify - JavaScript error: \(error.localizedDescription)")
            }
            
            if contentLength > 0 && !self.isContentLoaded && !self.isNotFoundNotified {
                // Content found!
                let foundMsg = "✅ Content FOUND! Length: \(contentLength)"
                print("\(TAG) \(foundMsg)")
                self.sendDebugLog(foundMsg)
                self.isContentLoaded = true
                
                UIView.animate(withDuration: 0.15) {
                    self._nativeWebView.alpha = 1.0
                }
                print("\(TAG) View made VISIBLE with animation")
                
                // Wait a bit for layout to complete before collecting dimensions
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) { [weak self] in
                    self?.collectWebViewInfoAndNotify(webView: webView)
                }
            } else if contentLength == 0 && retryCount < 5 {
                // No content yet, retry (max 5 times)
                let delay = 0.5 + Double(retryCount) * 0.2 // Increasing delay: 0.5s, 0.7s, 0.9s, 1.1s, 1.3s
                let retryMsg = "No content yet, will retry in \(delay)s (attempt \(retryCount + 1)/5)"
                print("\(TAG) \(retryMsg)")
                self.sendDebugLog(retryMsg)
                DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
                    if !self.isContentLoaded && !self.isNotFoundNotified {
                        self.checkContentAndNotify(webView: webView, retryCount: retryCount + 1)
                    }
                }
            } else if contentLength == 0 && retryCount >= 5 {
                // No content after all retries
                let giveUpMsg = "No content after \(retryCount + 1) attempts, giving up"
                print("\(TAG) \(giveUpMsg)")
                self.sendDebugLog(giveUpMsg)
                if !self.isContentLoaded && !self.isNotFoundNotified {
                    self.notifyContentNotFound()
                }
            }
        }
    }
    
    private func setupBoundsObserver() {
        print("\(TAG) Setting up bounds observer")
        
        // Observe bounds changes - when Flutter lays out the view
        boundsObserver = _nativeWebView.observe(\.bounds, options: [.new]) { [weak self] view, change in
            guard let self = self else { return }
            let newBounds = change.newValue ?? .zero
            


            
            // Initialize Dengage when view has non-zero size
            if newBounds.width > 0 && newBounds.height > 0 && !self.dengageInitialized {
                print("\(TAG) >>> View has size! Initializing Dengage...")
                self.initializeDengage()
            }
        }
        
        print("\(TAG) Bounds observer added")
    }
    
    private func initializeDengage() {
        guard !dengageInitialized else {
            print("\(TAG) initializeDengage - Already initialized, skipping")
            return
        }
        
        guard let propertyId = self.propertyId else {
            print("\(TAG) ERROR: propertyId is nil!")
            return
        }
        
        dengageInitialized = true
        
        
        
        let beforeState = "BEFORE SDK: isHidden=\(_nativeWebView.isHidden), bounds=\(_nativeWebView.bounds), alpha=\(_nativeWebView.alpha)"
        print("\(TAG) \(beforeState)")
        sendDebugLog(beforeState)
        
        Dengage.showInAppInLine(
            propertyID: propertyId,
            inAppInlineElement: _nativeWebView,
            screenName: screenName,
            customParams: customParams
        )
        
        let afterState = "AFTER SDK: isHidden=\(_nativeWebView.isHidden), bounds=\(_nativeWebView.bounds), alpha=\(_nativeWebView.alpha), subviews=\(_nativeWebView.subviews.count)"
        

        
        // THEN set up delegates (to override what SDK sets)
        // Wait for SDK to initialize and potentially load content
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
            guard let self = self else { return }
            let state05 = "0.5s AFTER SDK: isHidden=\(self._nativeWebView.isHidden), bounds=\(self._nativeWebView.bounds)"
            print("\(TAG) \(state05)")
            self.sendDebugLog(state05)
        }
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            guard let self = self else { return }
            
            self.setupWebViewDelegate()
            self.isInitialized = true
            
            // Log WebView state
            self.logWebViewState()
            
            // If page is already loaded, check content
            self.checkIfAlreadyLoaded()
            
            // Set up visibility observers AFTER initial loading attempt
            // This prevents false "notFound" during initialization
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) { [weak self] in
                self?.setupVisibilityObservers()
            }
        }
        
    }
    
    private func setupVisibilityObservers() {
        
        // Only observe isHidden - SDK explicitly hides view when no content
        visibilityObserver = _nativeWebView.observe(\.isHidden, options: [.new, .old]) { [weak self] view, change in
            guard let self = self else { return }
            let oldValue = change.oldValue ?? false
            let newValue = change.newValue ?? false
            
            
            // Only trigger if view becomes hidden AND we haven't loaded content yet
            if !oldValue && newValue && !self.isContentLoaded && !self.isNotFoundNotified {
                DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                    if !self.isContentLoaded && !self.isNotFoundNotified && self._nativeWebView.isHidden {
                        self.notifyContentNotFound()
                    }
                }
            }
        }
        
        // Note: We removed bounds observer as it was too aggressive
        // SDK sets bounds during initialization which triggered false "notFound"
        
    }
    
    private func checkAndNotifyNotFound() {
        
        
        guard !isContentLoaded && !isNotFoundNotified else {
            
            return
        }
        
        if _nativeWebView.isHidden {
            
            notifyContentNotFound()
        } else {
            
        }
    }
    
    private func sendDebugLog(_ message: String) {
        DispatchQueue.main.async { [weak self] in
            
            // Also send to Flutter console via print
            self?.methodChannel.invokeMethod("onDebugLog", arguments: ["message": message])
        }
    }
    
    private func notifyContentNotFound() {
        guard !isNotFoundNotified else {
            return
        }
        isNotFoundNotified = true
        
        let debugInfo = "CONTENT NOT FOUND - isHidden: \(_nativeWebView.isHidden), bounds: \(_nativeWebView.bounds), alpha: \(_nativeWebView.alpha)"
        
        sendDebugLog(debugInfo)
        
        methodChannel.invokeMethod("onContentNotFound", arguments: nil)
        print("\(TAG) onContentNotFound sent to Flutter")
    }
    
    private func setupWebViewDelegate() {
        print("\(TAG) Setting up WebView delegate")
        
        if let webView = findWKWebView(in: _nativeWebView) {
            webView.navigationDelegate = self
            currentWebView = webView
            
        } else {
            print("\(TAG) WKWebView NOT found, retrying in 0.1s")
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
                self?.retrySetupWebViewDelegate()
            }
        }
    }
    
    private func retrySetupWebViewDelegate() {
        print("\(TAG) retrySetupWebViewDelegate called")
        if let webView = findWKWebView(in: _nativeWebView) {
            webView.navigationDelegate = self
            currentWebView = webView
            
        } else {
            print("\(TAG) WKWebView still NOT found on retry")
        }
    }
    
    private func findWKWebView(in view: UIView) -> WKWebView? {
        if let webView = view as? WKWebView {
            return webView
        }
        for subview in view.subviews {
            if let webView = findWKWebView(in: subview) {
                return webView
            }
        }
        return nil
    }
    
    private func notifyContentLoaded(webView: WKWebView) {
        
        
        guard !isContentLoaded && !isNotFoundNotified else {
            
            return
        }
        isContentLoaded = true
        
        UIView.animate(withDuration: 0.15) {
            self._nativeWebView.alpha = 1.0
        }
        
        collectWebViewInfoAndNotify(webView: webView)
    }
    
    private func collectWebViewInfoAndNotify(webView: WKWebView) {
        print("\(TAG) collectWebViewInfoAndNotify - Starting")
        print("\(TAG) WebView URL: \(webView.url?.absoluteString ?? "nil")")
        print("\(TAG) WebView Title: \(webView.title ?? "nil")")
        print("\(TAG) WebView frame: \(webView.frame)")
        print("\(TAG) Native view frame: \(_nativeWebView.frame)")
        
        let nativeFrame = _nativeWebView.frame
        
        // If native frame is valid, use it
        if nativeFrame.width > 0 && nativeFrame.height > 0 {
            let contentHeight = Double(nativeFrame.height)
            let contentWidth = Double(nativeFrame.width)
            
            let webViewInfo: [String: Any?] = [
                "url": webView.url?.absoluteString,
                "title": webView.title,
                "contentHeight": contentHeight,
                "contentWidth": contentWidth,
                "canGoBack": webView.canGoBack,
                "canGoForward": webView.canGoForward
            ]
            
            
            self.methodChannel.invokeMethod("onContentLoaded", arguments: webViewInfo)
        } else {
            // Fallback to JavaScript if native frame is 0
            print("\(TAG) ⚠️ Native frame is 0x0, falling back to JavaScript")
            
            let script = """
                (function() {
                    var body = document.body;
                    var html = document.documentElement;
                    return {
                        height: Math.max(body.scrollHeight || 0, body.offsetHeight || 0, html.scrollHeight || 0, html.offsetHeight || 0, 150),
                        width: Math.max(body.scrollWidth || 0, body.offsetWidth || 0, html.scrollWidth || 0, html.offsetWidth || 0, 300)
                    };
                })();
            """
            
            webView.evaluateJavaScript(script) { [weak self] result, error in
                guard let self = self else { return }
                
                
                var contentHeight: Double = 150.0 // Default fallback
                var contentWidth: Double = 300.0  // Default fallback
                
                if let dimensions = result as? [String: Any] {
                    contentHeight = dimensions["height"] as? Double ?? 150.0
                    contentWidth = dimensions["width"] as? Double ?? 300.0
                }
                
                
                let webViewInfo: [String: Any?] = [
                    "url": webView.url?.absoluteString,
                    "title": webView.title,
                    "contentHeight": contentHeight,
                    "contentWidth": contentWidth,
                    "canGoBack": webView.canGoBack,
                    "canGoForward": webView.canGoForward
                ]
                
                
                self.methodChannel.invokeMethod("onContentLoaded", arguments: webViewInfo)
            }
        }
    }
    
    private func notifyContentError(error: Error?) {
        methodChannel.invokeMethod("onContentError", arguments: [
            "description": error?.localizedDescription ?? "Unknown error"
        ])
    }
}

// MARK: - WKNavigationDelegate
extension InAppinline: WKNavigationDelegate {
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        
    }
    
    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
        
    }
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
       
        // HTML içeriğini kontrol et
        if !isContentLoaded && !isNotFoundNotified {
            checkContentAndNotify(webView: webView)
        } else {
            print("\(TAG) >>> didFinish - SKIPPED (already loaded or notFound)")
        }
    }
    
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        
        notifyContentError(error: error)
    }
    
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        
        notifyContentError(error: error)
    }
    
    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        
        decisionHandler(.allow)
    }
}
